### DISCUSSION TITLE
Run Nextflow on a Docker container

### DISCUSSION THREAD

**Discussion by aarens66** [nextflow-io/nextflow#3782] [Q&A] [answered]
Hi,

I want to run Nextflow in a docker container because I want to avoid to maintain VM or run locally. Has anyone tried to create a container that is able to run nextflow and this calls a batch for the pipelines? for example, to run https://github.com/nextflow-io/rnaseq-nf?

Example: 

**run this on a container**: nextflow -c nextflow.config run https://github.com/nextflow-io/rnaseq-nf --outdir s3://nextflowbatch/nextflow_env/results

(Instead of running it in a virtual machine)

**Comment by mribeirodantas** [comment, marked as answer]
I assume that you will also want to run your tasks isolated from each other within containers. The first issue that arises from this is that it'll run a container within a container and this is prone to several issues and thus should not be done.

If you insist on running the Nextflow head job on a container, you can use the `-d` or `-dockerize` option from the CLI. But I warn you that there have even been discussions about deprecating this feature as it brings more issues than solutions.

**Reply by aarens66** [reply]
Hi, my idea was to run the processes in containers but in other instances using aws batch, this way the processes are not run inside the container where Nextflow executor is running. Instead of an EC2 Instance calling Batch my idea was to make it from a container calling Batch.

**Reply by mribeirodantas** [reply]
Oh, I see. In this case, I think it's OK.

**Reply by aarens66** [reply]
Thanks!

**Comment by byzantic** [comment]
Hi, I just opened an issue (https://github.com/nextflow-io/nextflow/issues/3811 ) that is related to this, however, I am a bit alarmed by the comment that there are problems with running containerized pipelines on a containerized nextflow server.

Could you point me to those other discussions please?

What I want to do is to have the nextflow server running as a docker container on a docker host, but also be able to run pipeline stages within other containers.

This is because the nextflow server may need to be launched on one or more docker hosts; I want to be able to scale out the number of server instances as required.

As for the pipeline stages, these may require very different environments / applications and it is not possible to provide all of them on the same platform, hence the need to provide containers for the pipeline stages.

**Reply by mribeirodantas** [reply]
[This is a blog post by the creator of "Docker-in-Docker"](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/).

**Reply by byzantic** [reply]
Hi,

Thanks for the reply. This is a good point, actually I suppose I should have said that I want to launch containerized pipeline stages from _within_ a container running nextflow, but not for the containers to be located within the nextflow container.

As the author of that paper points out, the way to do this is to bind the host /var/run/docker.sock into the container that runs the nextflow instance. I did attempt to do that when running the official nextflow docker image, but again, I couldnt get it to run the pipeline stages inside containers.

**Reply by mribeirodantas** [reply]
Oh, I see. I think the direction you're taking in the other issue (with `-dockerize` and help from @bentsherman) is the right way to go now 😄
